woocommerce/woocommerce-gutenberg-products-block
# Update ProductControl to use HOCs

This PR:
- Creates a `withProductVariations` HOC.
- Modifies `withSearchedProducts` HOC so it can be chained with other HOCs.
- Creates a `withSingleSelected` HOC, so components which accept a single-value `selected` prop, can use HOCs that expect an array for the `selected` prop.
- Updates the `<ProductControl />` component to use the new HOCs and show errors when they happen.

### Screenshots
![image](https://user-images.githubusercontent.com/3616980/65314905-77534000-db97-11e9-937a-ec5f1a9e4722.png)
_(`<ProductControl />` displaying an error)_

### How to test the changes in this Pull Request:

1. Create a post with a _Featured Product_ block.
2. Verify the Product Control works as expected and you can select variations as expected.
3. Modify [this line](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/components/utils/index.js#L70) in `assets/js/components/utils/index.js`:
```diff
-path: `${ ENDPOINTS.products }/${ productId }`,
+path: `wrong-path`,
```
4. Reload the previous post and verify the product control displays an error.
5. Undo the previous modification in the file and, instead, modify [this other line](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/974/files#diff-4af89e2259f067e12516f7bd47f82d2bR135):
```diff
-path: addQueryArgs( `${ ENDPOINTS.products }/${ product }/variations`, {
+path: addQueryArgs( `wrong-path`, {
```
6. Reload the post again and verify an error is displayed.
